From f51f69ac686f8450ca475e0794d87c084fea0703 Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <icenowy@aosc.xyz>
Date: Sat, 25 Feb 2017 16:27:51 +0800
Subject: [PATCH 10/66] arm64: dts: allwinner: add backlight control for
 Pinebook

Pinebook features a LCD backlight controlled by PWM.

Add support for it.

Signed-off-by: Icenowy Zheng <icenowy@aosc.xyz>
---
 .../arm64/boot/dts/allwinner/sun50i-a64-pinebook.dts | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/arch/arm64/boot/dts/allwinner/sun50i-a64-pinebook.dts b/arch/arm64/boot/dts/allwinner/sun50i-a64-pinebook.dts
index d7ee04014435..062f78da4818 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-a64-pinebook.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-a64-pinebook.dts
@@ -46,6 +46,7 @@
 
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/input/input.h>
+#include <dt-bindings/pwm/pwm.h>
 
 / {
 	model = "Pinebook";
@@ -55,6 +56,19 @@
 		serial0 = &uart0;
 	};
 
+	backlight: backlight {
+		compatible = "pwm-backlight";
+		pwms = <&pwm 0 50000 0>;
+		/*
+		 * Directly use PWM_POLARITY_INVERTED here cannot work,
+		 * and may even make backlight fail to be initialized
+		 * during U-Boot in next boot.
+		 */
+		brightness-levels = <100 70 60 50 40 30 20 10 0>;
+		default-brightness-level = <6>;
+		enable-gpios = <&pio 3 23 GPIO_ACTIVE_HIGH>; /* PD23 */
+	};
+
 	chosen {
 		stdout-path = "serial0:115200n8";
 	};
@@ -116,6 +130,12 @@
 	status = "okay";
 };
 
+&pwm {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pwm0_pins>;
+	status = "okay";
+};
+
 &r_rsb {
 	status = "okay";
 
-- 
2.13.6

